### PR TITLE
add SPDK vhost support

### PR DESCRIPTION
**Pull request by caiwenzh**

### CODE REVIEW

**Comment by liyimeng** [comment]
@caiwenzh This is a great effort!  I see you start to abstract the interface in [pkg/csi/server/routers.go](https://github.com/alibaba/open-local/pull/153#diff-f7c0f6c29920b9abf202223d8ad68533218cceed78eb31f912e022929c5739b7) in order to support more than one storage type.  Can we make this even more generic, in order to support more storage type, for example, zfs, the one I'm so longing fr :D   

@TheBeatles1994 should be take this opportunity to make some abstraction?

**Comment by liyimeng** [comment]
by the way, running go vendor is not a must, I would suggest you ignore vendor dir to keep this pr simple and easy to review. @caiwenzh  
When we get spdk landed and tested, go vendor can come in as as separated commit, what do u think?

**Review by TheBeatles1994 — COMMENTED** [review]
chart 包也适配下吧，比如 agent 要挂载宿主机哪些目录到容器中。关键信息可以提取到 values.yaml 中。

**Comment by TheBeatles1994** [comment]
> @TheBeatles1994 should be take this opportunity to make some abstraction?

@liyimeng Sure we will make some abstraction to support more storage type. We plan to do this in [scheduling-framework plugin](https://github.com/alibaba/open-local/pull/148/files) first, then migrate to scheduler extender, what do you think?

File: pkg/agent/discovery/discovery.go [hunk 1]
```diff
@@ -126,6 +145,10 @@ func (d *Discoverer) Discover() {
 		nlsCopy.Status.FilteredStorageInfo.UpdateStatus.LastUpdateTime = &lastUpdateTime
 		nlsCopy.Status.FilteredStorageInfo.UpdateStatus.Reason = ""
 
+		if d.spdk {
+			d.initSpdkBdevs(&nlsCopy.Status.FilteredStorageInfo.Devices)
```
[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/agent/discovery/discovery.go:149]: 每个block device 在SPDK中都需要有一个对应的bdev。这是在创建新发现的设备对应的bdev. 否则在SPDK是找不到这些设备的。
对于resourceToBeInited.vgs， 先检查对应的bdev是否已经创建，确保已经创建后，再创建lvstore (也就是VG)。

File: pkg/utils/spdk/spdk.go [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+}
+
+func NewSpdkClient(rpcSocket string) *SpdkClient {
+	if rpcSocket == "" {
+		rpcSocket = "/var/tmp/spdk.sock"
+	}
+
+	cli := SpdkClient{rpcSocket: rpcSocket}
+
+	cli.vhostUserStorePath = "/var/run/kata-containers/vhost-user/"
```
[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/utils/spdk/spdk.go:75]: okay.  还是统一作一个常量吧。 这些路径是kata里统一定义的，应该不会变的。

File: pkg/agent/discovery/discovery.go [hunk 3]
```diff
@@ -136,26 +159,98 @@ func (d *Discoverer) Discover() {
 	}
 }
 
+// Create AIO bdevs
+func (d *Discoverer) initSpdkBdevs(devices *[]string) {
+	var found bool
+
+	bdevs, err := d.spdkclient.GetBdevs()
+	if err != nil {
+		return
```
[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/agent/discovery/discovery.go:168]: okay

[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/agent/discovery/discovery.go:168]: okay. 在spdk.go中 API出错会打印相应的log。 而且会把调用这个API的函数打印出来。

File: pkg/agent/discovery/discovery.go [hunk 4]
```diff
@@ -136,26 +159,98 @@ func (d *Discoverer) Discover() {
...
+		for _, bdev := range *bdevs {
+			if bdev.GetFilename() == dev {
+				found = true
+				break
+			}
+		}
+		if !found {
+			bdevName := "bdev-aio" + strings.Replace(dev, "/", "_", -1)
+			if _, err := d.spdkclient.CreateBdev(bdevName, dev); err != nil {
+				log.Print("initSpdkBdevs - CreateBdev failed:", err.Error())
```
[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/agent/discovery/discovery.go:182]: okay。其他地方有的用的是go标准库里的log包，没有log.Info。 我统一改成用logrus吧

**Comment by caiwenzh** [comment]
> by the way, running go vendor is not a must, I would suggest you ignore vendor dir to keep this pr simple and easy to review. @caiwenzh When we get spdk landed and tested, go vendor can come in as as separated commit, what do u think?

okay

File: pkg/agent/discovery/discovery.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheBeatles1994) [reply pkg/agent/discovery/discovery.go:149]: nlsCopy.Status.FilteredStorageInfo.Devices 这里是筛选后的独占块设备，之后会作为独占盘给一整个 PV：https://github.com/alibaba/open-local/blob/main/docs/user-guide/type-device_zh_CN.md

**Comment by liyimeng** [comment]
> > @TheBeatles1994 should be take this opportunity to make some abstraction?
> 
> @liyimeng Sure we will make some abstraction to support more storage type. We plan to do this in [scheduling-framework plugin](https://github.com/alibaba/open-local/pull/148/files) first, then migrate to scheduler extender, what do you think?

Good plan!  @TheBeatles1994

File: pkg/agent/discovery/discovery.go [hunk 5]
```diff
@@ -126,6 +145,10 @@ func (d *Discoverer) Discover() {
 		nlsCopy.Status.FilteredStorageInfo.UpdateStatus.LastUpdateTime = &lastUpdateTime
 		nlsCopy.Status.FilteredStorageInfo.UpdateStatus.Reason = ""
 
+		if d.spdk {
```
[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/agent/discovery/discovery.go:148]: 在这个地方根据Discover发现和过滤后的结果创建对应的bdev 是比较合适的。如果放到InitResource() 中发现和过滤需要重新做一遍 。 函数名initSpdkBdevs 可能容易引起误解，将改成 createSpdkBdevs。

**Comment by peterzcst** [comment]
This PR has two features implemented if I'm not wrong, one for SPDK vhost support the other is for Kata block volume. Better to refactor this into two PRs

File: pkg/csi/nodeserver.go [hunk 6]
```diff
@@ -92,24 +102,35 @@ func newNodeServer(d *csicommon.CSIDriver, dName, nodeID, sysPath string) csi.No
...
 		client:               kubeClient,
+		localclient:          localclient,
 		driverName:           dName,
 		sysPath:              sysPath,
 		ephemeralVolumeStore: store,
 		inFlight:             NewInFlight(),
+		spdkSupported:        false,
 	}
+
+	go ns.checkSPDKSupport()
```
[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/csi/nodeserver.go:131]: 现在这样可以确保SPDK suppoting 的检查尽快执行并且避免遗忘 ，没有必要放到 (plugin *CSIPlugin) Run() 里。

File: pkg/csi/nodeserver.go [hunk 7]
```diff
@@ -148,6 +169,106 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
 
 	volCap := req.GetVolumeCapability()
+	if ns.spdkSupported {
+		var err error
+		device, exist := ns.ephemeralVolumeStore.GetDevice(volumeID)
+		if !exist || device == "" {
+			if volumeType != LvmVolumeType {
+				if sourceDevice, exists := req.VolumeContext[DeviceVolumeType]; exists {
+					bdevName := "bdev-aio" + strings.Replace(sourceDevice, "/", "_", -1)
+					if _, err := ns.spdkclient.CreateBdev(bdevName, sourceDevice); err != nil {
```
[the remark this answers is not in the document]
Comment (caiwenzh) [reply pkg/csi/nodeserver.go]: 在前面的代码里，会确保如果是临时卷就会自动将Device 类型转为LVM。 这里是为了避免如果是PV，如果之前没有创建好device, 这里确保device一定会被创建。

**Comment by TheBeatles1994** [comment]
lgtm

**Pull request merged**